### PR TITLE
nxos_vpc: pkl_vrf fixes for #57069

### LINKED ISSUE #57069: nxos_vpc breaks using default vrf
##### SUMMARY
When using pkl_vrf": "default" command is missing vrf value

##### ISSUE TYPE
- Bug Report

##### COMPONENT NAME
Module: nxos_vpc

##### ANSIBLE VERSION
```
ansible 2.7.2
  config file = /etc/ansible/ansible.cfg
  configured module search path = [u'/root/.ansible/plugins/modules', u'/usr/share/ansible/plugins/modules']
  ansible python module location = /usr/lib/python2.7/dist-packages/ansible
  executable location = /usr/bin/ansible
  python version = 2.7.15rc1 (default, Nov 12 2018, 14:31:15) [GCC 7.3.0]
```

##### CONFIGURATION
asterisk due privacy
```
CACHE_PLUGIN(/etc/ansible/ansible.cfg) = jsonfile
CACHE_PLUGIN_CONNECTION(/etc/ansible/ansible.cfg) = /**/config/ansible/facts
DEFAULT_HOST_LIST(/etc/ansible/ansible.cfg) = [u'/**/config/ansible/**/hosts.yml']
DISPLAY_SKIPPED_HOSTS(/etc/ansible/ansible.cfg) = True
HOST_KEY_CHECKING(/etc/ansible/ansible.cfg) = False
RETRY_FILES_ENABLED(/etc/ansible/ansible.cfg) = False
```

##### STEPS TO REPRODUCE
```
  nxos_vpc:
    domain: 10
    pkl_src: 1.1.1.2
    pkl_dest: 1.1.1.1
    pkl_vrf: default
```

##### EXPECTED RESULTS
```
    "commands": [
        "vpc domain 10",
        "peer-keepalive destination 1.1.1.1 source 1.1.1.2 vrf default",
```
##### ACTUAL RESULTS
```
    "commands": [
        "vpc domain 10",
        "peer-keepalive destination 1.1.1.1 source 1.1.1.2",
```

### PR DESCRIPTION
**Pull request by chrisvanheuveln**
##### SUMMARY
Fixes #57069

__Symptom__
When playbooks specify `pkl_vrf: default`, the result is that the cli does not set the `vrf` state.

__Analysis__
- First issue: `default` is a cli reserved word (for `peer-keepalive`) when used with the `vrf` keyword. It refers to the default rib.
    - This is confusing in several ways because `peer-keepalive`'s *default* vrf is the `management` vrf.

- Second issue: When changing only one optional value (e.g. `pkl_vrf`) while other optional values are idempotent (e.g. `pkl_src`), the result is that the idempotent values are ignored; unfortunately the device cli *replaces* the entire command, in which case the idempotent values are removed.
    - e.g. playbook specifies this:
     ```
     pkl_dest: 10.1.1.1
     pkl_src: 10.2.2.2
     pkl_vrf: my_vrf
     ```

     ```
     peer-keepalive dest 10.1.1.1 src 10.2.2.2             # original

     peer-keepalive dest 10.1.1.1 src 10.2.2.2 vrf my_vrf  # intended result

     peer-keepalive dest 10.1.1.1 vrf my_vrf               # actual result
     ```

- Third issue: the `pkl` getter was relying on positional data. This broke when the `udp` keyword nvgen'd where `vrf` used to appear (shifting all keywords to the right).

- Added new unit tests to validate `vrf` behavior

##### ISSUE TYPE
- Bugfix Pull Request

##### COMPONENT NAME
`nxos_vpc`

##### ADDITIONAL INFORMATION
- Tested on regression platforms: `N3K,N6k,N7K,N9K,N3K-F`

### CODE REVIEW

File: lib/ansible/modules/network/nxos/nxos_vpc.py [hunk 1]
```diff
@@ -262,19 +265,41 @@ def get_vpc(module):
...
+    pkl_existing = [i for i in existing.keys() if i.startswith('pkl')]
+
+    for pkl in pkl_existing:
+        param = module.params.get(pkl)
+        if not delta.get(pkl) and param and param == existing[pkl]:
+            # delta is missing this param because it's idempotent;
+            # however another pkl command has changed; therefore
+            # explicitly add it to delta so that the cli retains it.
+            delta[pkl] = existing[pkl]
+
```
Comment (chrisvanheuveln) [inline lib/ansible/modules/network/nxos/nxos_vpc.py:306]: Note: Mike wants to preserve "merge" behavior, so always retain existing source even if playbook does not specify.

Comment (chrisvanheuveln) [reply lib/ansible/modules/network/nxos/nxos_vpc.py:306]: latest commit now does merge.

**Comment by mikewiebe** [comment]
shipit

File: lib/ansible/modules/network/nxos/nxos_vpc.py [hunk 2]
```diff
@@ -58,9 +58,12 @@
   pkl_dest:
     description:
       - Destination (remote) IP address used for peer keepalive link
+      - pkl_dest is required whenever pkl options are used.
```
Comment (trishnaguha) [inline lib/ansible/modules/network/nxos/nxos_vpc.py:61]: Is `pkl_dest` required for both `pkl_vrf` and `pkl_src`?

Comment (chrisvanheuveln) [reply lib/ansible/modules/network/nxos/nxos_vpc.py:61]: Yes, `pkl_dest` is required for all `peer-keepalive` options:

```
n9k-109(config-vpc-domain)# peer-keepalive ?
  destination  Specify destination ip address of peer switch

n9k-109(config-vpc-domain)# peer-keepalive destination ?
  A.B.C.D  IPv4 address (A.B.C.D) of destination

n9k-109(config-vpc-domain)# peer-keepalive destination 192.168.1.1 ?
  <CR>
  hold-timeout  Hold timeout to ignore stale peer alive messages
  interval      Enter interval in milleseconds
  precedence    Precedence
  source        Source interface for hello
  tos           Type of Service
  tos-byte      Type of Service Byte
  udp-port      Enter UDP port number used for hello
  vrf           Vrf to be used for hello messages
```

There is also a dependency check for it here:
 https://github.com/ansible/ansible/blob/devel/lib/ansible/modules/network/nxos/nxos_vpc.py#L369

File: test/units/modules/network/nxos/test_nxos_vpc.py [hunk 3]
```diff
@@ -64,3 +80,82 @@ def test_nxos_vpc_present(self):
...
+            pkl_src='10.1.1.1',
+            pkl_vrf='test-vrf',
+        ))
+        self.execute_module(changed=True, device='_vrf_test', commands=[
+            'vpc domain 100',
+            'peer-keepalive destination 192.168.1.1 source 10.1.1.1 vrf test-vrf'
+        ])
+
+    def test_nxos_vpc_vrf_3(self):
+        # vrf 'my_vrf'-> vrf 'test-vrf'
```
Comment (mikewiebe) [inline test/units/modules/network/nxos/test_nxos_vpc.py]: `my_vrf -> obvious_vrf`

File: test/units/modules/network/nxos/test_nxos_vpc.py [hunk 4]
```diff
@@ -64,3 +80,82 @@ def test_nxos_vpc_present(self):
...
+        # vrf 'my_vrf' -> vrf 'my_vrf' (idempotence)
+        self.get_config.return_value = load_fixture('nxos_vpc', 'vrf_test_vpc_config')
+        set_module_args(dict(
+            domain=100,
+            pkl_dest='192.168.1.1',
+            pkl_src='10.1.1.1',
+            pkl_vrf='my_vrf',
+        ))
+        self.execute_module(changed=False, device='_vrf_test')
+
```
Comment (mikewiebe) [inline test/units/modules/network/nxos/test_nxos_vpc.py:150]: I think you should also add the following test

```python
    def test_nxos_vpc_vrf_7(self):
        # vrf 'my_vrf' -> vrf 'my_vrf' (idempotence)
        # pkl_src not in playbook but exists on device.
        self.get_config.return_value = load_fixture('nxos_vpc', 'vrf_test_vpc_config')
        set_module_args(dict(
            domain=100,
            pkl_dest='192.168.1.1',
            pkl_vrf='my_vrf',
        ))
        self.execute_module(changed=False, device='_vrf_test')
```

Comment (chrisvanheuveln) [reply test/units/modules/network/nxos/test_nxos_vpc.py:150]: done

File: test/units/modules/network/nxos/test_nxos_vpc.py (hunk 3 above)
Comment (chrisvanheuveln) [reply test/units/modules/network/nxos/test_nxos_vpc.py]: obviously

**Comment by mikewiebe** [comment]
shipit

File: lib/ansible/modules/network/nxos/nxos_vpc.py (hunk 2 above)
Comment (trishnaguha) [reply lib/ansible/modules/network/nxos/nxos_vpc.py:61]: @chrisvanheuveln Are these three arguments required together?
In that case you can use `required_together=[['pkl_dest', 'pkl_vrf', 'pkl_src']]` and pass this to AnsibleModule class?

Comment (chrisvanheuveln) [reply lib/ansible/modules/network/nxos/nxos_vpc.py:61]: @trishnaguha `pkl_dest` is the top-level keyword so it's a dependency but the other arguments are all optional; e.g. if my playbook specifies `pkl_vrf` then I must include `pkl_dest`, but `pkl_src` is still optional.

**Review by trishnaguha — APPROVED** [review]

**Pull request merged**